feat: implement TODOs & fix tests

cmltaWt0 opened the pull request:
I'll continue the work done by @andrii-hantkovskyi in this PR.

## Prev Description

This PR addresses several previously noted TODOs in the `migrate_from_modulestore` task logic and improves testing accuracy and coverage.

## Key Changes

- Implemented incremental creation of ModulestoreBlockSource and ModulestoreBlockMigration records during migration.
- Updated logic to resolve `target_ids` for collection population based on created migration records.
- Refactored slug generation for block keys to avoid collisions and ensure compatibility.
- Improved error handling when creating component version content to gracefully handle duplicates.
- Cleaned up obsolete inline comments and removed unnecessary legacy code paths.
- Updated test expectations for `slugify_source_usage_key()` to reflect hashed values.
- Fixed failing test for `migrate_from_modulestore_success_course` to use assertions properly.

Commit:
feat: implement TODOs & fix tests

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -533,36 +505,94 @@ def _migrate_component(
         )
+
+    # Add hash suffix to reduce collision probability
+    key_hash = hashlib.md5(str(key).encode()).hexdigest()[:8]
```
Won't this make it so components are always created new instead of replaced/skipped? For example, if I import ablock with the id`block-v1:org+course+run+type@html+block@blah`, the first time it will end up as a component with the key:

* `org__course__run__blah_1234abcd`

and the second time it'll end up with the key:
* `org__course__run__blah_5678efgh`

thus creating two different components.

cmltaWt0 replied:
The md5 algorithm is deterministic and should produce the same hash output for the same input (am I missing something?).
It works in tests and should actually work in theory. But I'm wondering is it a solution to a collision issue and do we really need this additional hash.

kdmccormick replied:
> The md5 algorithm is deterministic and should produce the same hash output for the same input (am I missing something?).

Nope, you're right, my bad

> But I'm wondering is it a solution to a collision issue and do we really need this additional hash.

It would make it _a little_ less likely that imported items would conflict with usage keys of existing target library blocks. But, that probability is already low, because the structure of the imported slugs (without the additional md5 hash) is `MyOrg__MyCourse__MyRun__MyBlockId`, and it would be surprising if any target library blocks just coincidentally had the same slug as that.

The issues with these approaches (both with and without the additional md5 hash) are:

* We cannot *entirely* guarantee that a target library block will not collide using this method, because it's possible for the existing blocks in the target library to have _any_ slug, regardless of how convoluted we make the target slug format.
* The target library block slugs become very long. For example, when I imported the demo course, all the slugs were like `OpenedX__DemoX__DemoCourse__<hash>`, whereas it would have been generally fine for them to just be `<hash>`, as the source hash is already highly likely to be unique.

Here's the solution I can think of that would solve both of those issues:

* Before importing the components and containers, grab a list of existing source-to-target mappings from the ModulestoreBlockMigration table, put it in a dict called something like `existing_source_to_target_keys`.
* When creating any given component or container:
  * see if the source usage key is in `existing_source_to_target_keys`
    * If so, use that target key. This will ensure that re-imported blocks always map to the same target key.
    * If not, then see if the source block's slug already in use in the library.
      * If it's free, then use that slug.
      * If it's in use, then append `_1`, `_2`, etc. until a unique slug is found

This would both ensure that imported slug are always unique, while also avoiding having to encode the entire source usage key (including org/course/run) into the target slug.

I'm going to bring this up today at the library sync to make sure that it's an agreeable path forward.

kdmccormick replied:
FYI @ormsbee ^ , in case you have a chance to pre-read this before the sync, no worries if not

cmltaWt0 replied:
Agreed. I like the proposed approach.

kdmccormick replied:
From today's meeting: https://github.com/openedx/frontend-app-authoring/issues/1681#issuecomment-3119568678

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -533,36 +505,94 @@ def _migrate_component(
+    status: UserTaskStatus
+) -> None:
+    """
+    Create ModulestoreBlockSource and ModulestoreBlockMigration objects incrementally.
```
This does create the ModulestoreBlockSource and ModulestoreMigration artifacts one-by-one, but it only does so after the actual structure (the Component and Container instances) have been imported, which takes a a few minutes for large courses. The benefit of creating them incrementally would be so that we can see migration progress during those few minutes *as the the structure is being imported*. If that is not performant/possible, then I think it is best to just create the ModulestoreBlockSource and ModulestoreMigration artifacts in-bulk at the end.

cmltaWt0 replied:
Implemented the bulk version for comparison.
Code became a bit complicated though.

Let's choose a better option and I'll amend the commit.

Performance:
- incremental for 39 Blocks 0.06 sec
- bulk for 39 Blocks 0.006 sec

cmltaWt0 replied:
Disabled for now since I've encountered an issue with adding to a collection - the blocks are not visible on the add_to_collection step when using the bulk version.

cmltaWt0 committed:
feat: implement _create_migration_artifacts_in_bulk

Left the incremental implementation to compare.

TODO:
- remove one of the competitive versions
- discuss the better option
- amend this commit and clear commin history

Performance:
- incremental for 39 Blocks 0.06 sec
- bulk for 39 Blocks 0.006 sec

cmltaWt0 committed:
feat: implement Handling duplicates in modulestore migrator

TODO:
- remove extensive loggins, it's currently required
  for debugging purpuse
- Update tests regarding the new slugification logic
  and according to new field
- update API
- debug and fix performance issue during authoring_api.add_to_collection
- refactor existing_source_to_target_keys by moving it and
  some common arguments from _migrate_node to MigrationContext
- test API and remove unused fields

cmltaWt0 committed:
feat: implement Containers restoring

Changes:
- refactor slugification
- restore a deleted container

cmltaWt0 committed:
feat: implement Containers deduplication

TODO:
- decrease code duplication
- add tests

cmltaWt0 committed:
test: fix modulestore_migrator test

Changes:
- fix updating the existing_source_to_target_keys dictionary process
- fix new slug generation for Contaners

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -212,67 +216,51 @@ def migrate_from_modulestore(
+        # target_package_id:
+        # replace_existing:
+        # preserve_url_slugs:
+    # This will reduce parameter count from 10+ to 2-3
```
As long as we're talking about a `frozen=True` dataclass, that sounds good to me @cmltaWt0 .

cmltaWt0 replied:
> As long as we're talking about a `frozen=True` dataclass, that sounds good to me @cmltaWt0 .

Absolutely!

cmltaWt0 replied:
Implemented.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -291,16 +279,27 @@ def migrate_from_modulestore(
         )
+        if block_target_pks:
+            log.info("Started adding entities to a collection")
+            # @@TODO: Investigate an issue with huge
+            # time consumption during this step
+            # For existing and deleted blocks this step takes
+            # up to 110 seconds
+            # For new blocks this step takes up to 3 seconds
```
FYI, I would be OK merging with this part performing slowly as long as we create a follow-up issue to optimize it and link it here.

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -525,44 +570,375 @@ def _migrate_component(
+    title: str,
+    existing_source_to_target_keys: dict[UsageKey, PublishableEntityVersion],
+    preserve_url_slugs: bool = False
+) -> LibraryUsageLocatorV2:
```
```suggestion
) -> LibraryContainerLocator:
```

Unlike with modulestore, containers are not actually XBlocks in Learning Core, so they don't have usage keys. The LibraryContainerLocator is a separate class.

cmltaWt0 replied:
Sure, it's just a typo - the returned value is indeed a **LibraryContainerLocator**.
Fixed

kdmccormick commented on cms/djangoapps/modulestore_migrator/tasks.py:
```diff
@@ -525,44 +570,375 @@ def _migrate_component(
+        publishable_entity.key for _, publishable_entity in
+        existing_source_to_target_keys.items()
+        if publishable_entity.key.startswith(base_key)
+    )
```
I think this could be computed once at the beginning of the modulestore_migrator and passed around with the MigrationContext object so that we don't have to re-run the query for every imported block.

cmltaWt0 replied:
It could but I've found a use case when we have several units/sections/subsections in one course with the same title. For this use case (and when preserve_slugs=False) it's critical to re-fetch previously added PEs.

For this use case we are adding newly added publishable entities to the `existing_source_to_target_keys` [here](https://github.com/kdmccormick/edx-platform/pull/35/files#diff-7991f5d644e0e6a3946ce330599ce9249e4b86d58809bbe45cce445a6c88df35R434) to avoid the work on stale data. This is a WIP and I want to come up with a clear understanding whether we need to use both `existing_keys` and  `existing_pe_keys` or not.

kdmccormick replied:
> This is a WIP and I want to come up with a clear understanding whether we need to use both existing_keys and existing_pe_keys or not.

Sounds good. I think it should be possible just to use existing_pe_keys and avoid the database query.

cmltaWt0 committed:
feat: implement MigrationContext

Changes:
- refactor migration context usage
- update tests
- add new tests
- fix typo in return type for _deduplicate_container

cmltaWt0 committed:
refactor: avoid redundant DB queries

cmltaWt0 committed:
fix: Rest API create action returns incorrect value

cmltaWt0 committed:
refactor: update Rest API contract

Changes:
- remove `name` field
- rename `modulestoremigration` to `parameters`
- update docstring to correctly describe `state_text`
